foxai/explainer/computer_vision/algorithm/xrai.py

```diff
@@ -0,0 +1,687 @@
+        """
+        output_attr: np.ndarray = -np.inf * np.ones(
+            shape=attribute_sample.shape, dtype=float
+        )
+        current_area_perc: float = 0.0
+        current_mask = np.zeros(attribute_sample.shape, dtype=bool)
+
+        masks_trace: List[Tuple[np.ndarray, float]] = []
+        remaining_masks = dict(enumerate(segments))
+
+        added_masks_cnt: int = 1
+        # While the mask area is less than area_th and remaining_masks is not empty
+        while current_area_perc <= area_perc_th:
+            best_gain: float = -np.inf
+            best_key: int
+            remove_key_queue: List[int] = []
+            for mask_key in remaining_masks:
+                mask: np.ndarray = remaining_masks[mask_key]
+                # If mask does not add more than min_pixel_diff to current mask, remove
+                mask_pixel_diff = _get_diff_cnt(mask, current_mask)
+                if mask_pixel_diff < min_pixel_diff:
+                    remove_key_queue.append(mask_key)
+                    continue
+                gain: float = gain_fun(mask, attribute_sample, current_mask)
+                if gain > best_gain:
+                    best_gain = gain
+                    best_key = mask_key
+            for key in remove_key_queue:
+                del remaining_masks[key]
+
+                # if dictionary is empty end processing
+            if not remaining_masks:
+                break
+
+            added_mask = remaining_masks[best_key]
+            mask_diff = _get_diff_mask(added_mask, current_mask)
+            masks_trace.append((mask_diff, best_gain))
+
+            current_mask = np.logical_or(current_mask, added_mask)
+            current_area_perc = np.mean(current_mask)
+            output_attr[mask_diff] = best_gain
+            del remaining_masks[best_key]  # delete used key
+            added_masks_cnt += 1
+
+        uncomputed_mask: np.ndarray = output_attr == -np.inf
+        # Assign the uncomputed areas a value such that sum is same as ig
+        output_attr[uncomputed_mask] = gain_fun(uncomputed_mask, attribute_sample, None)
+        masks_trace_array_list = [
+            v[0] for v in sorted(masks_trace, key=lambda x: -x[1])
+        ]
+        if np.any(uncomputed_mask):
+            masks_trace_array_list.append(uncomputed_mask)
+
+        final_mask_trace: List[np.ndarray] = masks_trace_array_list
+        if integer_segments:
+            integer_segment_mask = np.zeros(shape=attribute_sample.shape, dtype=int)
+            for i, mask in enumerate(masks_trace_array_list):
+                integer_segment_mask[mask] = i + 1
+
+            final_mask_trace = [integer_segment_mask]
+        return output_attr, final_mask_trace
+
+
+class XRAICVExplainer(Explainer):
+    """XRAI algorithm explainer."""
+
+    def calculate_features(
+        self,
+        model: ModelType,
+        input_data: torch.Tensor,
+        pred_label_idx: TargetType = None,
```

Comment:
Please use Optional[X] if None is allowed.

Reply:
I see that all explanation methods have this issue. I will create another PR with typing-related issues.